src/server/frontend_wayland/foreign_toplevel_manager_v1.cpp

```diff
@@ -0,0 +1,696 @@
+    void send_state(MirWindowFocusState focused, MirWindowState state);
+
+    /// Sends the .closed event and makes this surface inert
+    void should_close();
+
+private:
+    /// Modifies the surface if possible, silently fails if not
+    void attempt_modify_surface(shell::SurfaceSpecification const& spec);
+
+    /// Wayland requests
+    ///@{
+    void set_maximized();
+    void unset_maximized();
+    void set_minimized();
+    void unset_minimized();
+    void activate(struct wl_resource* seat);
+    void close();
+    void set_rectangle(struct wl_resource* surface, int32_t x, int32_t y, int32_t width, int32_t height);
+    void destroy();
+    void set_fullscreen(std::experimental::optional<struct wl_resource*> const& output);
+    void unset_fullscreen();
+    ///@}
+
+    std::weak_ptr<shell::Shell> const weak_shell;
+    std::weak_ptr<scene::Surface> weak_surface;
+
+    /// Used to choose the state to unminimize/unfullscreen to
+    ///@{
+    MirWindowState cached_normal_state{mir_window_state_restored}; ///< always either restored or a maximized state
+    bool cached_fullscreen{false};
+    ///@}
+};
+
+}
+}
+
+auto mf::create_foreign_toplevel_manager_v1(
+    wl_display* display,
+    std::shared_ptr<shell::Shell> const& shell,
+    std::shared_ptr<Executor> const& wayland_executor,
+    std::shared_ptr<SurfaceStack> const& surface_stack) -> std::shared_ptr<ForeignToplevelManagerV1Global>
+{
+    return std::make_shared<ForeignToplevelManagerV1Global>(display, shell, wayland_executor, surface_stack);
+}
+
+// ForeignToplevelManagerV1Global
+
+mf::ForeignToplevelManagerV1Global::ForeignToplevelManagerV1Global(
+    wl_display* display,
+    std::shared_ptr<shell::Shell> const& shell,
+    std::shared_ptr<Executor> const& wayland_executor,
+    std::shared_ptr<SurfaceStack> const& surface_stack)
+    : Global{display, Version<2>()},
+      shell{shell},
+      wayland_executor{wayland_executor},
+      surface_stack{surface_stack}
+{
+}
+
+void mf::ForeignToplevelManagerV1Global::bind(wl_resource* new_resource)
+{
+    new ForeignToplevelManagerV1{new_resource, *this};
+}
+
+// ForeignSceneObserver
+
+mf::ForeignSceneObserver::ForeignSceneObserver(
+    std::shared_ptr<Executor> const& wayland_executor,
+    ForeignToplevelManagerV1* manager)
+    : wayland_executor{wayland_executor},
+      manager{manager}
+{
+}
+
+mf::ForeignSceneObserver::~ForeignSceneObserver()
+{
+    clear_surface_observers();
+}
+
+void mf::ForeignSceneObserver::surface_added(std::shared_ptr<scene::Surface> const& surface)
+{
+    create_surface_observer(surface);
+}
+
+void mf::ForeignSceneObserver::surface_removed(std::shared_ptr<scene::Surface> const& surface)
+{
+    std::lock_guard<std::mutex> lock{mutex};
+    auto const iter = surface_observers.find(surface);
+    if (iter == surface_observers.end())
+    {
+        log_error("Can not remove ForeignSurfaceObserver: surface %p not in observers map", surface.get());
```

Comment:
Non blocking, but is logging `surface.get()` useful? (Maybe we need some standard approach to identifying surfaces in the log: Name (if set), client session etc.)